File: packages/distance/src/edit/bag.ts
/**
 * Bag distance — fast approximation of edit distance.
 *
 * bag(a, b) = max(|a|, |b|) - |a ∩ b|
 * where |a ∩ b| is the multiset intersection size.
 *
 * This is a lower bound on Levenshtein distance and is much faster to compute.
 *
 * Time: O(m + n)
 */

import {
  normalize,
  buildCharFreqArray,
  CHAR_FREQ_SIZE,
  charFrequencyMap,
  intersectCount,
} from "../utils";

const _freqA = new Int32Array(CHAR_FREQ_SIZE);
const _freqB = new Int32Array(CHAR_FREQ_SIZE);

/**
 * Compute the bag distance between two strings.
 *
 * @param a - First string
 * @param b - Second string
 * @returns Bag distance (non-negative integer)
 */
export function bagDistance(a: string, b: string): number {
  let intersection: number;

  // ASCII fast path
  _freqA.fill(0);
  _freqB.fill(0);
  if (buildCharFreqArray(_freqA, a) && buildCharFreqArray(_freqB, b)) {
    intersection = 0;
    for (let i = 0; i < CHAR_FREQ_SIZE; i++) {
      intersection += Math.min(_freqA[i], _freqB[i]);
    }
  } else {
    intersection = intersectCount(charFrequencyMap(a), charFrequencyMap(b));
  }

  return Math.max(a.length, b.length) - intersection;
}

/**
 * Compute the normalized bag distance similarity in [0, 1].
 *
 * @param a - First string
 * @param b - Second string
 * @returns Similarity score where 1 means identical
 */
export function bagDistanceNormalized(a: string, b: string): number {
  return normalize(bagDistance(a, b), Math.max(a.length, b.length));
}

File: packages/distance-wasm/src/edit/bag.rs
/// Bag distance — fast approximation of edit distance.
///
/// bag(a, b) = max(|a|, |b|) - |a ∩ b|
/// Time: O(m + n)

use crate::utils::{normalize, build_char_freq, intersect_count};

pub fn bag_distance(a: &str, b: &str) -> u32 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();

    let mut freq_a = [0u32; 128];
    let mut freq_b = [0u32; 128];

    let intersection = if build_char_freq(&mut freq_a, a_bytes) && build_char_freq(&mut freq_b, b_bytes) {
        intersect_count(&freq_a, &freq_b)
    } else {
        // Non-ASCII fallback: simple byte multiset
        let mut fa = std::collections::HashMap::<u8, u32>::new();
        let mut fb = std::collections::HashMap::<u8, u32>::new();
        for &b in a_bytes { *fa.entry(b).or_insert(0) += 1; }
        for &b in b_bytes { *fb.entry(b).or_insert(0) += 1; }
        let mut count = 0u32;
        for (k, &v) in &fa {
            count += v.min(*fb.get(k).unwrap_or(&0));
        }
        count
    };

    (a_bytes.len().max(b_bytes.len()) as u32) - intersection
}

pub fn bag_distance_normalized(a: &str, b: &str) -> f64 {
    let max_len = a.len().max(b.len()) as u32;
    normalize(bag_distance(a, b), max_len)
}
